### PR TITLE
Runtime support to take advantage of Modular PlusCal

### PR DESCRIPTION
**Pull request by rmascarenhas**
One of the main advantages of having Modular PlusCal is being able to have direct communication channels between PlusCal processes in the implementation while having the spec use global variables as usual.

This PR introduces what is the first step towards providing runtime support for these distributed channels. It implements the interface we discussed in #95  for both:

* **Global Variables**. This just wraps the existing global state management API.
* **TCP Channels**: This implements direct communication across two nodes; it should be significantly more efficient (although no numbers were collected) than relying on global state that could live anywhere, moves around and could be under high contention.

### TCP Channels run on top of RPC

This is more of a comment than a discussion point: TCP channels are implemented on top of Go's RPC implementation. This is because each PlusCal process is already using an IP/port combination with RPC for the existing synchronization/bookkeeping/global state communication. I didn't want to have the user specify yet another port just for this, so we are using RPC instead of TCP sockets for the channel.

### Maintaining collections of channels

Currently I have a working example of a distributed single producer/single consumer queue on top of TCP channels (source code in the Appendix). However, it is still unclear how we are going to compile specs that make use of a _collection_ of channels. This is a feature being used in most of the Modular PlusCal specs we wrote and yet I think we failed to consider this case. I updated `examples/mpcal/dqueue.tla` spec in this pull request to illustrate the issue, but here are the relevant sections:

```tla
archetype AProducer(ref network, stream)
 variables requester; {
    p: while (TRUE) {
        requester := network;
        network[requester] := stream;
    }
}

  fair process (Consumer \in 1..NUM_CONSUMERS) == instance AConsumer(network[self], ref processor)
      mapping network via LossyNetwork;
  fair process (Producer = NUM_CONSUMERS+1) == instance AProducer(ref network, ref stream)
      mapping network via LossyNetwork
      mapping stream via CyclicReads;
```
The difficulties with the spec above are:

* `instance AConsumer(network[self],`: It is going to be hard to inject the "correct" channel here without human intervention.

* Ideally, I would want my "producer" archetype to get a _collection_ of channels to each consumer. Being able to do so makes it easy to model sending a message to a particular consumer or even broadcasting -- not needed in this example, but I remember broadcasting was the motivating example behind `ref` parameters in procedures.

TL;DR: not entirely clear how this "collection" use-case will map to the Go interface we proposed.

### How does the developer chooses what is the underlying implementation of archetype params?

Some options:

* Have the developer change the source code PGo generates
    - What does PGo generate when run then? Non-compiling code?

* Have the developer specify before compilation what the channels are (global vs TCP or UDP, etc)
   - Could be done via annotations or using the existing compilation configuration file.

## Appendix

Here is the source code for a single producer/single consumer distributed queue using this TCP channel API and the `ArchetypeResource` interface:

```go
package main

import (
	"fmt"
	"math/rand"
	"os"
	"pgo/distsys"
	"strconv"
)

var queue []string
var processName string
var processArgument string
var err error
var globalState *distsys.StateServer

func init() {
	queue = []string{}
	if len(os.Args) != 3 {
		fmt.Printf("Usage: %s processName(processArgument) ip:port\n", os.Args[0])
		os.Exit(1)
	}
	processPlusArgument := os.Args[1]
	ipPort := os.Args[2]
	processName, processArgument = distsys.ParseProcessId(processPlusArgument)
	peers := []string{"127.0.0.1:2222", "127.0.0.1:3333"}
	coordinator := peers[0]
	globalState, err = distsys.NewStateServer(processPlusArgument, peers, ipPort, coordinator, map[string]interface{}{
		"queue": queue,
	}, distsys.NewRandomMigrate(ipPort))
	if err != nil {
		panic(err)
	}
}

func Producer(self int, network distsys.ArchetypeResource) {
	for {
		if !true {
			break
		}

		// wait for request from consumer
		tmpRequester, err := network.Acquire(distsys.READ_ACCESS)
		if err != nil {
			panic(err)
		}

		requester := tmpRequester.(int)
		fmt.Printf("Producer: got request from %d\n", requester)

		// send some "data" for the consumer to consume
		if err := network.Commit(rand.Intn(100)); err != nil {
			panic(err)
		}
	}
}

func Consumer(self int, network distsys.ArchetypeResource) {
	for {
		if !true {
			break
		}

		// request more data from producer
		if err := network.Commit(self); err != nil {
			panic(err)
		}

		tmpData, err := network.Acquire(distsys.READ_ACCESS)
		if err != nil {
			panic(err)
		}

		data := tmpData.(int)
		fmt.Printf("Consumer: processing %d\n", data)
	}
}

func main() {
	err = globalState.WaitPeers()
	chanName := "prod-cons"
	var network *distsys.TCPChannel

	if err != nil {
		panic(err)
	}

	switch processName {
	case "Producer":
		i, err0 := strconv.Atoi(processArgument)
		if err0 != nil {
			panic("Process Producer requires an integer argument; err = " + err0.Error())
		}
		network, err = distsys.NewTCPChannel(chanName, "Consumer(0)", globalState.PeerConnections(), 1024)
		if err != nil {
			panic(err)
		}

		Producer(i, network)
	case "Consumer":
		i, err0 := strconv.Atoi(processArgument)
		if err0 != nil {
			panic("Process Consumer requires an integer argument; err = " + err0.Error())
		}
		network, err = distsys.NewTCPChannel(chanName, "Producer(1)", globalState.PeerConnections(), 1024)
		if err != nil {
			panic(err)
		}

		Consumer(i, network)
	default:
		panic("Unknown process " + processName)
	}
	err = globalState.WaitPeers()
	if err != nil {
		panic(err)
	}
}
```

### CODE REVIEW

**Comment by fhackett** [comment]
@rmascarenhas This is an interesting PR - a lot of the fundamentals look good. It's pretty cool to see the stuff we've been debating all this time actually somewhat implemented. I have one issue and some responses to the questions you've raised.

My issue: that `init` function will try to open sockets, parse command-line arguments, etc on import of the Go module. I don't think that is a good experience for someone trying to reuse `Consumer` or `Producer` from another module - in fact, most of those globals should probably just be local to `main` since they are only accessed there. The only exception to this should be non-archetype processes, but since those should only exist for "demo" purposes I think we can be very lazy about that kind of thing. (just put `go (func(){ self := /* something */; arg1 := /* something else */; /* the process body compiled */ })()` in the main function for example, since a non-archetype is non-reusable anyway)

> How does the developer chooses what is the underlying implementation of archetype params?

Wasn't the idea to let the user specify the variable implementations at run-time by having their own main function that passed different parameters to the archetypes? If we do that we can be really lazy and just assume the global state API, since the code we generate (outside of the archetype body) is just kind of a "demo" that will be overridden by one or more "production" main functions.

> Maintaining collections of channels

You raise an interesting point - I had forgotten about that issue.

If we stick to the idea of "optimisations are the implementation's problem" we could provide a library utility that might fix the issue at least partially. What if create something like a `ShardedArchetypeResource` that implements a generic key-value dictionary (that is, only works if you assign a key-value slice to it) and proxies assignments to a group of ArchetypeResource instances?

Then we could transparently shard anything we want without needing to alter codegen. For example in the situation of single broadcaster multiple receivers we could give the broadcaster one of these sharded resources then distribute the individual resources to each of the individual receivers.

As for how PGo might generate a default instance of this, that's a bit trickier. Maybe we could detect this by looking for the most nested `ref a.b.c["d"]...` parameter and generating nested ShardedArchetypeResource parameters accordingly?

Here are some informal descriptions of the Read, Write and Abort algorithms (where S is an ordered map from keys to ArchetypeResource):

```
Read(S, permissions):
  if permissions & READ:
    result := allocate or retrieve cached slice
  else:
    result := nil
  for k, v in S:
    tmp := v.Read(permissions)
    if permissions & READ:
      result[k] = tmp
  return result
```

```
Write(S, newValue):
  cached := retrieve cached previous values
  for k, v in newValue:
    if k not in S: error, we don't know who to send it to
    if cached[k] != newValue[k]:
      S[k].Write(v)
    else:
      S[k].Abort() // or require ArchetypeResource to optimise same-value assignments
```

```
Abort(S):
  for _, v in S:
    v.Abort()
```

Let me know what you think - as usual this is just me thinking about the issue for a little while, so I may have made some kind of mistake/bad assumption/etc somewhere.

**Comment by rmascarenhas** [comment]
> My issue: that init function will try to open sockets, parse command-line arguments, etc on import of the Go module. I don't think that is a good experience for someone trying to reuse Consumer or Producer from another module - in fact, most of those globals should probably just be local to main since they are only accessed there.

The `init` function was introduced some time ago in the backend (I guess around the time of the migration to the new AST). I don't see a problem with it for two reasons: first, we do need those connections anyway for the internal barrier algorithm (see [wiki](https://github.com/UBC-NSS/pgo/wiki/Synchronization-Barrier-Protocol)); second, we haven't discussed in detail how reuse of other archetypes across specs is going to work (both in terms of PlusCal compilation as well as Go compilation). I'd rather have us focused to implement what is specified in the [current proposal](https://github.com/UBC-NSS/pgo/wiki/Modular-PlusCal) and worry about the other things once we have that.

> Wasn't the idea to let the user specify the variable implementations at run-time by having their own main function that passed different parameters to the archetypes? If we do that we can be really lazy and just assume the global state API, since the code we generate (outside of the archetype body) is just kind of a "demo" that will be overridden by one or more "production" main functions.

Makes sense, I guess I was thinking in terms of "how can we minimize the amount of generated code the user has to change?", but I guess as long as we strive to make the archetypes immutable, it should be good enough for the time being.

A comment on "let's default to global state API when compiling": that was my initial conclusion too; and that makes me realize that for that to be possible, all channels should be unidirectional. My current TCP channel implementation is bidirectional and would break that abstraction. I'll fix that.

> If we stick to the idea of "optimisations are the implementation's problem" we could provide a library utility that might fix the issue at least partially. What if create something like a ShardedArchetypeResource that implements a generic key-value dictionary (that is, only works if you assign a key-value slice to it) and proxies assignments to a group of ArchetypeResource instances?

Yes, that's also what I thought at first, but wanted to ask everyone to see what the opinions were. Thanks for drafting the algorithms, they make sense. I'll also work on this to prototype how it would look like.

**Comment by rmascarenhas** [comment]
Some more food for thought.

Consumer spec:

```tla
archetype AConsumer(ref conn, ref processor) {
    c: while (TRUE) {
        conn := self;
        processor := conn;          
    }
}
```

The idea here is that the consumer sends its own identifier to the producer in order to request more data. Then, it processes whatever is read from the connection (i.e., whatever piece of data the producer sends back).

This works under the assumption of a TCP connection; however, the abstraction breaks if we swap the implementation to global variables (in that implementation, reading back from the "connection" could return the value of `self` just sent).

**Comment by fhackett** [comment]
@rmascarenhas most of that makes a lot of sense.

To add a couple of things:

- `init` was introduced by me as a means of initialising static data in a general way (Go constant definition rules are ad-hoc in the least helpful way...). The only time it was reasonable to pack globals/mutable state in there was when global scope was actually useful, i.e vanilla PCal compilation. For archetype compilation I consider it counter to the purpose of archetypes, since the whole point was being able to say "because we have archetypes you can easily use MPCal code from other Go modules with no modifications, just write this setup code and call the archetype". I'm not saying to scrap the connections/state/etc, but that we should move all that (except the constant definitions) to the main function where it belongs.

- Re: things that assume a TCP connection, good point. You've run into a potential user issue there - it can be easy to get used to a particular implementation of your variables and forget the "one true semantics" (since the model checker sees globals, deviating from that would harm your correctness argument). It may also be worth asking under what conditions your TCP implementation is correct, since blocking until you receive something is not correct global variable behaviour. For full correctness, maybe it's worth having a version of the TCP channel that latches the last value sent and returns that if someone queries "too soon" instead of blocking as it seems your implementation does currently?

**Comment by rmascarenhas** [comment]
> I'm not saying to scrap the connections/state/etc, but that we should move all that (except the constant definitions) to the main function where it belongs.

I guess I see what you're saying -- I'll keep this in mind when working on the Go backend for MPCal. :+1:

> For full correctness, maybe it's worth having a version of the TCP channel that latches the last value sent and returns that if someone queries "too soon" instead of blocking as it seems your implementation does currently?

Yes, that would be an alternative, although one that would force both specs and code to do something awkward. If possible we should really strive to have "real" message passing.

I modified the specification to have this notion of "unidirectionality" of writes (and therefore having TCP channels provide the same semantics as a global variable):

```tla
(***************************************************************************
--mpcal DistQueue {
  mapping macro NetworkReads {      
      read {
          await Len($variable) > 0;
          with (msg = Head($variable)) {
              $variable := Tail($variable);
              yield msg;
          };
      }
      
      write { assert false; }
  }
  
  (* variable is expected to be mapped to a collection of message queues.
     Reading from it will return the first message from *some* non-empty queue *)
  mapping macro NetworkSelectRead {
      read {
          with (ch \in { c \in DOMAIN $variable : Len($variable[c]) > 0 }) {
              with (msg = Head($variable[ch])) {
                  $variable[ch] := Tail($variable[ch]);
                  yield msg;
              };
          };
      }
      
      write { assert false; }
  }
  
    mapping macro LossyNetworkWrites {      
      read { assert false; }
      
      write {
          either {
              yield $variable
          } or {
              await Len($variable) < BUFFER_SIZE;
              yield Append($variable, $value);
          }
      }
  }
  
  mapping macro CyclicReads {
      read  {
          $variable := ($variable + 1) % BUFFER_SIZE;
          yield $variable;
       
      }
      
      write { yield $variable }
  }
    
  (* consumer: Processes one element read from the network at a time, infinitely *)
  archetype AConsumer(prducerIn, ref producerOut, ref processor) {
      c: while (TRUE) {
          c1: producerOut := self;     (* request data by sending its own identifier *)
          c2: processor := producerIn; (* read data coming from the producer *)
      }
  }
  
  archetype AProducer(consumerIns, ref consumerOuts, stream)
  variables requester; {
      p: while (TRUE) {
          p1: requester := consumerIns;          (* read from the first  consumer that requests data *)
          p2: consumerOuts[requester] := stream; (* send back data from the stream (implementation specific *)
      }
  }
  
  variables consumerReadEnd = [c \in 1..NUM_CONSUMERS |-> <<>>],
            consumerWriteEnd = [c \in 1..NUM_CONSUMERS |-> <<>>],
            queueProcessor = 0,
            queueStream = 0;
            
  fair process (Consumer \in 1..NUM_CONSUMERS) == instance AConsumer(consumerReadEnd[self], ref consumerWriteEnd[self], ref queueProcessor)
      mapping consumerReadEnd via NetworkReads
      mapping consumerWriteEnd via LossyNetworkWrites;
  fair process (Producer = NUM_CONSUMERS+1) == instance AProducer(consumerWriteEnd, ref consumerReadEnd, ref queueStream)
      mapping consumerReadEnd via LossyNetworkWrites
      mapping consumerWriteEnd via NetworkSelectRead
      mapping queueStream via CyclicReads;
}
***************************************************************************)
```

Notes

* I separated the mapping macros in `Read` and `Write` parts mostly for sanity checks (i.e., to make sure the spec writer would accidentally try to read from the consumer's write end of the pipe). In practice, the separation shouldn't be strictly necessary if coupled with "careful" use of global variables.

* This "idiom" may potentially work well for both specs and implementation, at the expense of requiring the user to do a bit of plumbing.

**Comment by rmascarenhas** [comment]
Given our latest meeting, this pull request should be good to go (we still need to support channel collections as briefly discussed here).
